File: p1/1345.JumpGameIV.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var minJumps = function(arr) {
  if (arr.length <= 1) {
    return 0;
  }

  if (arr.length == 2) {
    return 1;
  }

  const last = arr.length - 1;
  if (arr[0] === arr[last]) {
    return 1;
  }

  let ans = 0;
  let stack = [0];
  let val, pos, nextJumps, nextPos;

  const jumps = {};
  const visited = new Set();
  visited.add(0);

  for (let i = 0; i < arr.length; i++) {
    val = arr[i];

    if (jumps[val]) {
      jumps[val].push(i);
    } else {
      jumps[val] = [i];
    }
  }

  while (stack && stack.length > 0) {
    const next = [];

    for (let i = 0; i < stack.length; i++) {
      pos = stack[i];
      if (pos === last-1) {
        return ans + 1;
      }

      if (pos > 0 && !visited.has(pos-1)) {
        next.push(pos-1);
        visited.add(pos-1);
      }

      if (pos < last-1 && !visited.has(pos+1)) {
        if (pos === last-2) {
          return ans + 2;
        }

        next.push(pos+1);
        visited.add(pos+1);
      }

      val = arr[pos];
      nextJumps = jumps[val];

      for (let j = 0; j < nextJumps.length; j++) {
        nextPos = nextJumps[j];

        if (nextPos === last) {
          return ans + 1;
        }

        if (nextPos === last - 1) {
          return ans + 2;
        }

        if (visited.has(nextPos)) {
          continue;
        }

        next.push(nextPos);
        visited.add(nextPos);
      }
    }

    stack = next;
    ans++
  }

  return ans;
};

File: challenges/1499/1345.JumpGameIV.py
'''
Given an array of integers arr, you are initially positioned at the first index of the array.

In one step you can jump from index i to index:

i + 1 where: i + 1 < arr.length.
i - 1 where: i - 1 >= 0.
j where: arr[i] == arr[j] and i != j.
Return the minimum number of steps to reach the last index of the array.

Notice that you can not jump outside of the array at any time.

Example 1:

Input: arr = [100,-23,-23,404,100,23,23,23,3,404]
Output: 3
Explanation: You need three jumps from index 0 --> 4 --> 3 --> 9. Note that index 9 is the last index of the array.
Example 2:

Input: arr = [7]
Output: 0
Explanation: Start index is the last index. You do not need to jump.
Example 3:

Input: arr = [7,6,9,6,9,6,9,7]
Output: 1
Explanation: You can jump directly from index 0 to index 7 which is last index of the array.

Constraints:

1 <= arr.length <= 5 * 10^4
-10^8 <= arr[i] <= 10^8
'''


from typing import List
from collections import defaultdict


class Solution:
  def minJumps(self, arr: List[int]) -> int:
    group = defaultdict(list)
    for i, val in enumerate(arr):
      group[val].append(i)
      
    curr, nxt = set([0]), set()
    seen = set([0])
    steps = 0
    end = len(arr)-1
    # print(group)
    
    while curr:
      if end in curr:
        return steps
      
      for pos in curr:
        if pos-1 not in seen and pos-1 >= 0:
          nxt.add(pos-1)
          seen.add(pos-1)
          
        if pos+1 not in seen and pos+1 <= end:
          if pos+1 == end:
            return steps+1
          
          nxt.add(pos+1)
          seen.add(pos+1)
          
        val = arr[pos]
        for nxt_pos in group[val]:
          if nxt_pos not in seen:
            if nxt_pos == end:
              return steps + 1
            
            nxt.add(nxt_pos)
            seen.add(nxt_pos)
        
        group.pop(val, None)
      
      curr, nxt = nxt, curr
      nxt.clear()
      steps += 1
      
    return end
  